msg/RadarReturn.msg

```diff
@@ -7,4 +7,4 @@ float32 azimuth                          # Angle (in radians) in the azimuth pla
 float32 elevation                        # Angle (in radians) in the elevation plane between the sensor and the detected return.
                                          # Negative angles are below the sensor. For 2D radar, this will be 0.
 float32 doppler_velocity                 # The doppler speed (m/s) of the return.
-float32 amplitude                        # The amplitude of the return (dB).
+float32 amplitude                        # The radar cross section of the return (RCS linear/m^2).
```

Comment:
I would generally reject a unilateral change in units of a ROS message without much further discussion from other users of the message. Please motivate this change in detail and "@" copy the other contributors to this repository for review

Reply:
This are 2 changes that we can also separate in the discussion.

- The current definition is missing what the value of `amplitude` relates to.
  dB only specified the scale, not the unit.
  Hence it is under-specified and would result in a sensor specific scale.
  It would be better to describe the scene, and not the sensor-specific amplitude value.
- The second change is, that we would propose to change the scale from logarithmic to linear.
  Since the sensor would measure it linear and most (or all?) calculations would also require a linear value.

To calculate RCS a reference measurement of an object with a known RCS is required.
For example a metal ball, where the RCS is equal to the geometric cross section.
(Hence it can also be done if the manufacturer doesn't provide this value.)
It can then be calculated by
```
RCS = (intensity[measurement] / intensity[referenceObject]) * (distance[referenceObject]^2 / distance[measurement]^2) * CrossSectionalArea[referenceObject]
```

Reply:
@aaronfultonnz, @fred-apex-ai, @icolwell-as and @radarAaron please feel free to contribute to this discussion.

Reply:
@tobiasneumann I'm not an expert here, so I don't provide a conclusion and invite What is @radarAaron @icolwell-as @aaronfultonnz device-specific comments

> The current definition is missing what the value of amplitude relates to.
dB only specified the scale, not the unit.

What @tobiasneumann suggests would also require renaming the field from `amplitude` to something like `rcs`. This would be a breaking change. An inconvenience to any consumer of this package, but useful to signal that something significant has changed. If you want to change the meaning of a field through a comment, the compiler will not flag the change for legacy uses and you risk having a legacy publisher and a new subscriber disagreeing on the meaning of a field, which we should avoid. 

> The second change is, that we would propose to change the scale from logarithmic to linear.
Since the sensor would measure it linear and most (or all?) calculations would also require a linear value.

 I just want to add that for the one radar sensor I worked with, the device returns a cluster RCS with unit dBm².  Given that, the logarithmic scale seems more natural to describe this sensor's output.  

I personally don't know if it's more useful to have a value on the log or linear for consumers of this message

Reply:
We could instead have a new message type that uses the more modern information if that would make sense instead!

Reply:
For major changes to fields and meanings, having a new type would help avoid confusion on the user side

Reply:
That seems like a good solution to me. @tobiasneumann can you make a new message with this change? Perhaps `RadarRCSReturn`? 

That should be an easy merge

Reply:
I try to group all your replies and go into the different points that have been made.

> What @tobiasneumann suggests would also require renaming the field from `amplitude` to something like `rcs`. This would be a breaking change. An inconvenience to any consumer of this package, but useful to signal that something significant has changed. If you want to change the meaning of a field through a comment, the compiler will not flag the change for legacy uses and you risk having a legacy publisher and a new subscriber disagreeing on the meaning of a field, which we should avoid.

That is a good point, if we want to go this way, the field name should be renamed.

>  * Delphi/Aptiv MRR & ESR: These radars are quite old now, but still fairly common in R&D and consumer vehicles. They provide the simple dB field, which was the [original inspiration](https://github.com/astuff/astuff_sensor_msgs/blob/melodic/radar_msgs/msg/RadarDetection.msg#L10) for the unit of this `amplitude` field. These radars do not provide RCS.
>
>  [...]
> Aside from the API change pains, I do understand the need for a better field with clearer units. Would introducing a new field and leaving the existing `amplitude` field only for legacy use be an option?

Just as a side note, RCS can be calculated with a reference measurement.
Hence it is not required to be provided by the manufacturer/driver directly.

> I just want to add that for the one radar sensor I worked with, the device returns a cluster RCS with unit dBm². Given that, the logarithmic scale seems more natural to describe this sensor's output.
> 
> I personally don't know if it's more useful to have a value on the log or linear for consumers of this message

This is actually a good point for (one of) the proposed changes.
Because `dBm²` is not the unit that is currently defined in the `amplitude` field.
Could it be that this sensor, already provided its data as RCS, just with a logarithmic scale?
If the field would be used with the currently described unit `dB`, it would be a sensor specific value and in algorithms it could only be used to distinguish measurements from a stream relativity to previous measurements.
(or with context knowledge of a specific solution setup, where the used sensors are known, but I don't think this use case should be taken into account for message definition).

> For major changes to fields and meanings, having a new type would help avoid confusion on the user side
> [...]
> Perhaps `RadarRCSReturn`?

I don't like this so much, because one of the powers of ROS are the defined messages, that enable the algorithms to be implemented interdependent from specific solutions/setups/sensors.
Because of the current definition, this is already broken for different sensors.
But introducing different messages, would result in drivers and algorithms to could have not matching interfaces.
Furthermore I would expect (stating that, without knowing it) that for current implementations, the value used to fill the `amplitude` field is not equal along different sensors.
And hence can only be used in solutions where the algorithms are implemented and/or configured for the used sensors (which works, but see my point above about these kind of setups, for this discussion).
(also please correct me if my assumption about this is not correct).

Taking all this in consideration I would propose a different solution.
- Changing the field name from `amplitude` to `rcs`.
- Using a logarithmic scale (`dBm^2`), instead of the currently proposed linear one.

My reasoning for this is.
- It would produce a visible breaking change for users of this package.
- The scale stays logarithmic (changing the scale to linear, is of much smaller importance, than the other proposed change)
- For solutions that are filling the field with wrong values or require a solution specific configuration, it doesn't matter if they have to fill a different field with the wrong value.
  (and because of my arguments above, this should cover all current implementations)

If you oppose this, the, in my opinion, next best solution would be to add a bool (or enum) field to the message, that states whether the value in the field is following the old `dB` unit/definition, or the new one.
I think that would enable us to not diverge with messages, but only require very small changes for existing.

However, I think having kind of 2 different definitions in one msg is quite uncommon in the ROS context.
(hence, only my 2. suggestion.)
But I think it would still be better than having 2 messages, with slide differences.
This would just be my 3. prefered option.

Reply:
2 Points regarding this.
- ROS has some mechanism to warn about different Versions of messages.
  But I only know about that it exists, not what exactly is been checked.
- This is just 1 out of 4 changes, that are been requested.
  I would think it would be best to discuss all of them, before a new version of this package is been released (but there I don't know the workflow for ROS)
  And the other changes are less complicated/dangerous, but change the fields more extensively.

Reply:
@tobiasneumann You are right, if we introduce a breaking change, we better look at all changes holistically. @SteveMacenski has asked for a new maintainer on https://discourse.ros.org/t/radar-messages-call-for-maintenance-help/34560. If your PRs can wait a bit, I suggest we wait a bit more and see who wants to take over maintenance, then roll out a new version with the new maintenance team while @SteveMacenski is still involved to have a proper handover.

Reply:
Hello everyone,
we would support the following proposal from @tobiasneumann 
> 
> Taking all this in consideration I would propose a different solution.
> 
> * Changing the field name from `amplitude` to `rcs`.
> * Using a logarithmic scale (`dBm^2`), instead of the currently proposed linear one.
>

Reply:
I've adapted the MR to :point_up:  change.
Does this resolve this discussion?